### DISCUSSION TITLE
Error:  torch.linalg.cholesky: For batch 0: U(X,X) is zero, singular U.. Backward pass will not work

### DISCUSSION THREAD

**Discussion by EcustBoy** [facebookresearch/theseus#429] [Q&A]
## ❓ Questions and Help
Hi author~ I try to use theseus in my motion planner model, and I design some type of cost based on **th.AutoDiffCostFunction**, and add them to objective, but when I start run it, during gradient backward step, the following type of error will occur: 

**There was an error while running the linear optimizer. Original error message: torch.linalg.cholesky: For batch 0: U(X,X) is zero, singular U.. Backward pass will not work. To obtain the best solution seen before the error, run with torch.no_grad()**
(X means number unfixed number, e.g. 2)

I wanna what's the possible reason

**Comment by luisenp** [comment]
Hi @EcustBoy. This looks like you have a zero jacobian matrix, which possible means that you have some cost weights set to zero, or that your error is zero at that point. Do you have some code to reproduce this error?

**Comment by EcustBoy** [comment]
> Hi @EcustBoy. This looks like you have a zero jacobian matrix, which possible means that you have some cost weights set to zero, or that your error is zero at that point. Do you have some code to reproduce this error?

Thanks for your reply~ I later debugged step by step, as you said, found that when some specific cost term were sparse (exists some zero value or even all zero value), such error would be reproduced.

Then I try the following several methods:

(1) directly **remove the specific sparse cost term**, and make sure **the rest cost terms are dense**(no zero value)

(2) r**edesign the specific sparse cost term to dense** form(no zero value)

And I found **method (1) still doesn't solve the problem**, the same error still exists,  while **method (2) can solve it**. 
(I'm sure  all cost weights are non zero value)

so my problems are:
-  I'm not clear why method (1) is invalid
-  Although method (2) is valid,  **in planner design some cost items are inherently sparse**, I don't think it's a good idea to manually modify the original cost formulation


I'll thanks much for your more advice ~

**Comment by luisenp** [comment]
After your changes in 1), are you getting exactly the same error? Or is it a singular matrix, but not necessarily zero?

**Comment by EcustBoy** [comment]
> After your changes in 1), are you getting exactly the same error? Or is it a singular matrix, but not necessarily zero?

Thanks for your quick reply~ 
- not exactly the same error I remember (e.g.  the "X" is different in error info "..., U(X,X) is zero, ..."  ).
-  It's possible I think, but I don't know how to check the values of the matrix， I just call the interface directly

**Comment by luisenp** [comment]
Is it possible for you to share a script that reproduces this error? There can be many reasons why this error occurs, it's hard to offer advice w/o knowing more details about your use case.

**Comment by luisenp** [comment]
Hi @EcustBoy. Any updates on this? Can we close this issue?

**Comment by EcustBoy** [comment]
> Hi @EcustBoy. Any updates on this? Can we close this issue?

Hi author~  I change the original variable from vehicle trajectory position to vehicle control (i.e. accelaration and steering rate), so the original cost function formation also changed,  and I guess it helps avoid jacobian matrix to be irreversible.  now it seems my motion planner runs normally now during training.  

But actually I don't know exactly the theoretical explanation for such case :-)  

Finally I also wanna know whether the theseus optimizer design specific trick to avoid such case? (e.g. add some tiny units matrix to original jacobian matrix)

Thank you very much for your patient reply~~  can you take time to answer my above doubts or give some suggestions finally? then the issue can be closed I think~

**Comment by luisenp** [comment]
Hi @EcustBoy. Were you using our motion planning code when you started getting this error? Sorry I can't offer an explanation yet, since I don't have the full details of what you were doing. 

With regards to the optimizer, you can use [LevenbergMarquardt](https://github.com/facebookresearch/theseus/blob/main/theseus/optimizer/nonlinear/levenberg_marquardt.py#L114) which adds some damping terms to the hessian approximation to account for ill-conditioned systems. The valid `optimizer_kwargs` are the keywords in `compute_delta` and `_complete_step`. The "ellipsoidal" option means that you add the hessian's diagonal times the damping factor; otherwise you add an identity. And "adaptive" means that the damping will be changed over the course of optimization depending on how the error is decreasing. 

Let me know if you have any other questions.
